R/join_data.R

```diff
@@ -1,90 +1,38 @@
+#' @seealso \code{\link{code}}, \code{\link[dplyr]{mutate-joins}},
+#'          \code{\link[dplyr]{filter-joins}}
+#' @author Stephen Su
 #' @export
-joindata <- function(.data, imported_data,
-                     origin_join_col, import_join_col,
-                     join_method, left, right) {
-
-    mc <- match.call()
-    dataname <- mc$.data
-    importname <- mc$imported_data
-
-    for (i in 1:length(origin_join_col)) {
-        if (((origin_join_col[i] == "") &
-             (import_join_col[i] != "")) |
-            ((origin_join_col[i] != "") &
-             (import_join_col[i] == ""))) {
-            stop(paste(
-              "Must select at least one column",
-              "from each dataset to match on"
-            ))
-        }
+join_data <- function(data_l, data_r, by = NULL,
+                      how = c("inner", "left", "right", "full", "anti", "semi"),
+                      suffix_l = ".x", suffix_r = ".y") {
+    expr <- rlang::enexprs(data_l, data_r)
+    how <- rlang::arg_match(how)
+    join_fn <- rlang::parse_expr(sprintf("dplyr::%s_join", how))
+    fn_arg <- rlang::list2(expr[[2]], by = rlang::enexpr(by))
+    if (how %in% c("inner", "left", "right", "full")) {
+        fn_arg <- c(fn_arg, suffix = rlang::expr(c(!!suffix_l, !!suffix_r)))
     }
-
-    col_names <- ""
-    for (i in seq_along(origin_join_col)) {
-        if ((origin_join_col[i] != "") & (import_join_col[i] != "")) {
-            col_names <-
-                paste0(
-                    col_names,
-                    "'",
-                    origin_join_col[i],
-                    "'='",
-                    import_join_col[i],
-                    "',"
-                )
-        }
-    }
-    col_names <- substr(col_names, 1, nchar(col_names) - 1)
-
-    byfml <- ""
-    if (col_names != "") {
-        byfml <- sprintf(", by = c(%s)", col_names)
-    }
-
-    if (join_method %in% c("left_join", "right_join", "full_join")) {
-        suf <- paste0(", suffix = c('.", left, "', '.", right, "')")
-    } else {
-        suf <- ""
-    }
-
-    exp <- ~.DATA %>%
-        .FUN(.IMP.BY.SUFFIX)
-
-    exp <- replaceVars(exp,
-        .DATA = dataname,
-        .IMP = importname,
-        .BY = byfml,
-        .FUN = paste("dplyr", join_method, sep = "::"),
-        .SUFFIX = suf
-    )
-
-
-    if (all(origin_join_col == "") & all(import_join_col == "")) {
-        res <- suppressMessages(interpolate(exp))
-        vars <- capture.output(
-            tmp <- dplyr::inner_join(.data, imported_data),
-            type = "message"
-        )
-        rm(tmp)
-        origin_join_col <-
-            eval(parse(text = gsub(".+ = ", "", vars)))
-        import_join_col <- origin_join_col
+    expr <- rlang::expr(!!expr[[1]] %>% (!!join_fn)(!!!fn_arg))
+    if (is.null(by)) {
+        by <- capture.output(data <- eval_code(expr), type = "message") |>
+            paste(collapse = " ") |>
+            stringr::str_remove("Joining, by =") |>
```

Comment:
In dplyr 1.1.0, the message produced by `dplyr::*_join()` functions was changed which breaks the `capture.output()` code. Although it's an easy fix for me this time, would this imply a source of inconsistency in the future as the message from dplyr join functions might change again in the future?